packages/labs/nextjs/src/index.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type {NextConfig} from 'next';
+
+// TODO(augustjk) Consider adding option for adding template shadowroot polyfill
+// automatically
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface LitSsrPluginOptions {}
+
+export = (_pluginOptions: LitSsrPluginOptions = {}) =>
+  (nextConfig: NextConfig = {}) => {
+    return Object.assign({}, nextConfig, {
+      webpack: (config, options) => {
+        const {isServer, nextRuntime, webpack} = options;
+        config.module.rules.unshift({
+          // Grab entry points for all pages
+          // TODO(augustjk) This may not work for the new "app" directory
+          // https://github.com/lit/lit/issues/3657
+          test: /\/pages\/.*\.(?:jsx?|tsx?)$/,
+          // Exclude Next's own distributed files as they're commonjs and won't
+          // play nice with `imports-loader`
+          exclude: /next\/dist\//,
+          loader: 'imports-loader',
+          options: {
+            // This adds a side-effectful import which monkey patches
+            // `React.createElement` in the server and imports
+            // `lit/experimental-hydrate-support.js` in the client.
+            imports: ['side-effects @lit-labs/ssr-react/enable-lit-ssr.js'],
+          },
+        });
+
+        if (isServer && nextRuntime === 'nodejs') {
+          // Next.js uses this externals setting to skip webpack bundling
+          // external code for server environments but we specifically want to
+          // catch react/jsx-runtime imports for external packages as well.
+          // This might change in a future update to Next.js.
+          // This is what we're grabbing:
+          // https://github.com/vercel/next.js/blob/412dfc52cc5e378e4f74e44af698dad25031a938/packages/next/src/build/webpack-config.ts#L1378-L1429
+          const nextHandleExternals = config.externals[0];
```

Comment:
Since this is/was an array, is it safe to only call `nextHandleExternals`? Will it always be the one-element array referenced in the link above?

Reply:
Next could choose to change their default webpack config but for now it's a single element array. If the user provides their own enhancement of the webpack config, or composes with another plugin that does, that could be problematic as I describe in the comment lower in the code. It's brittle but I think hard to anticipate that kind of composition so I think we'll have to address that if we get reports of that.

Reply:
Could you file an issue with Next on how to do this in a safe way? Maybe they could provide a hook for overriding/wrapping handleExternals.

Reply:
Done!